common/cf-guest/src/client_impls.rs

```diff
@@ -369,26 +394,95 @@ impl<PK: PubKey> ClientState<PK> {
+        header: Header<PK>,
     ) -> Result<bool> {
-        todo!()
+        fn get_timestamp_ns<T, E>(state: Option<T>) -> Result<Option<u64>>
+        where
+            E: ToString,
+            T: TryInto<ConsensusState, Error = E>,
+        {
+            match state.map(|state| state.try_into()) {
+                None => Ok(None),
+                Some(Ok(state)) => Ok(Some(state.timestamp_ns.get())),
+                Some(Err(err)) => Err(error(err)),
+            }
+        }
+
+        let height = header.block_header.block_height;
+        let height = ibc::Height::new(1, height.into())?;
+
+        Ok(match ctx.consensus_state_neighbourhood(client_id, height)? {
+            Neighbourhood::This(state) => {
+                // If we already have existing consensus for given height, check
+                // that what we’ve been sent is the same thing we have.  If it
+                // isn’t, that’s evidence of misbehaviour.
+                let existing_state = state.try_into().map_err(error)?;
+                let header_state = ConsensusState::from(&header);
+                existing_state != header_state
+            }
+
+            Neighbourhood::Neighbours(prev, next) => {
+                // Otherwise, make sure that timestamp of each consensus is
+                // strictly increasing.  If it isn’t, that’s evidence of
+                // misbehaviour.
+                let header_time_ns = header.block_header.timestamp_ns.get();
+                if let Some(prev_time_ns) = get_timestamp_ns(prev)? {
+                    if header_time_ns <= prev_time_ns {
+                        return Ok(true);
+                    }
+                }
+                if let Some(next_time_ns) = get_timestamp_ns(next)? {
+                    if header_time_ns >= next_time_ns {
+                        return Ok(true);
+                    }
+                }
+                false
+            }
+        })
     }
 
-    fn check_for_misbehaviour_misbehavior(
+    fn check_for_misbehaviour_in_misbehavior(
         &self,
-        _ctx: &impl guestchain::Verifier<PK>,
-        _misbehaviour: Misbehaviour<PK>,
+        _ctx: &impl CommonContext,
+        _client_id: &ibc::ClientId,
+        misbehaviour: Misbehaviour<PK>,
     ) -> Result<bool> {
-        todo!()
+        use core::cmp::Ordering;
+
+        let Misbehaviour { header1, header2 } = misbehaviour;
+
+        // If blocks have the same height they must be the same (i.e. have the
+        // same hash).  Otherwise, check if their timestamp is strictly
+        // increasing.  If those conditions are not met we have a proof of
+        // misbehaviour.
+        let block1 = header1.block_header;
+        let block2 = header2.block_header;
+        Ok(match block1.block_height.cmp(&block2.block_height) {
+            Ordering::Less => block1.timestamp_ns >= block2.timestamp_ns,
+            Ordering::Equal => header1.block_hash != header2.block_hash,
```

Comment:
==?

Reply:
The condition here checks if the header is invalid.  That is, if the two block hashes aren’t equal, than there is proof of misbehaviour.